src/core/SampleBuffer.cpp

```diff
@@ -869,122 +431,109 @@ bool SampleBuffer::play(
-		}
+		const auto available = end - index;
+		std::copy_n(m_data.begin() + index, available, out.begin());
+		std::fill_n(out.begin() + available, frames - available, sampleFrame{});
+		return out;
 	}
 
-	*tmp = MM_ALLOC<sampleFrame>( frames);
+	auto numFramesCopied = *backwards ?
+		std::min(frames, index - loopStart) :
+		std::min(frames, loopEnd - index);
 
-	if (loopMode == LoopOff)
-	{
-		f_cnt_t available = end - index;
-		memcpy(*tmp, m_data + index, available * BYTES_PER_FRAME);
-		memset(*tmp + available, 0, (frames - available) * BYTES_PER_FRAME);
-	}
-	else if (loopMode == LoopOn)
+	auto updateBackwards = [&]()
 	{
-		f_cnt_t copied = qMin(frames, loopEnd - index);
-		memcpy(*tmp, m_data + index, copied * BYTES_PER_FRAME);
-		f_cnt_t loopFrames = loopEnd - loopStart;
-		while (copied < frames)
+		if (!*backwards && index + numFramesCopied >= loopEnd)
 		{
-			f_cnt_t todo = qMin(frames - copied, loopFrames);
-			memcpy(*tmp + copied, m_data + loopStart, todo * BYTES_PER_FRAME);
-			copied += todo;
+			*backwards = true;
 		}
-	}
-	else
+		else if (*backwards && index - numFramesCopied == loopStart)
+		{
+			*backwards = false;
+		}
+	};
+
+	if (loopMode == LoopMode::LoopPingPong)
 	{
-		f_cnt_t pos = index;
-		bool currentBackwards = pos < loopStart
-			? false
-			: *backwards;
-		f_cnt_t copied = 0;
+		updateBackwards();
+	}
 
+	*backwards ?
+		std::copy_n(m_data.rbegin() + loopEnd - index, numFramesCopied, out.begin()) :
+		std::copy_n(m_data.begin() + index, numFramesCopied, out.begin());
 
-		if (currentBackwards)
-		{
-			copied = qMin(frames, pos - loopStart);
-			for (int i = 0; i < copied; i++)
-			{
-				(*tmp)[i][0] = m_data[pos - i][0];
-				(*tmp)[i][1] = m_data[pos - i][1];
-			}
-			pos -= copied;
-			if (pos == loopStart) { currentBackwards = false; }
-		}
-		else
+	const auto loopFrames = loopEnd - loopStart;
+	while (numFramesCopied < frames)
+	{
+		if (loopMode == LoopMode::LoopPingPong)
 		{
-			copied = qMin(frames, loopEnd - pos);
-			memcpy(*tmp, m_data + pos, copied * BYTES_PER_FRAME);
-			pos += copied;
-			if (pos == loopEnd) { currentBackwards = true; }
+			updateBackwards();
 		}
 
-		while (copied < frames)
-		{
-			if (currentBackwards)
-			{
-				f_cnt_t todo = qMin(frames - copied, pos - loopStart);
-				for (int i = 0; i < todo; i++)
-				{
-					(*tmp)[copied + i][0] = m_data[pos - i][0];
-					(*tmp)[copied + i][1] = m_data[pos - i][1];
-				}
-				pos -= todo;
-				copied += todo;
-				if (pos <= loopStart) { currentBackwards = false; }
-			}
-			else
-			{
-				f_cnt_t todo = qMin(frames - copied, loopEnd - pos);
-				memcpy(*tmp + copied, m_data + pos, todo * BYTES_PER_FRAME);
-				pos += todo;
-				copied += todo;
-				if (pos >= loopEnd) { currentBackwards = true; }
-			}
-		}
-		*backwards = currentBackwards;
+		auto framesToCopy = std::min(frames - numFramesCopied, loopFrames);
+		*backwards ?
+			std::copy_n(m_data.rbegin() + loopEnd - index, framesToCopy, out.begin() + numFramesCopied) :
```

Comment:
I think `index` isn't supposed to appear here since the segment before looping is already handled above.

Reply:
Yeah, it seems I messed up there. This is simply the wrong calculation. This would work if ``loopEnd`` was equal to ``m_data.size()``, but that obviously may not be the case.

To illustrate (for clarity and understanding), assuming ``loopEnd = 800``, ``index = 700``, and ``m_data.size() = 1000``,
That would make ``m_data.rbegin() + loopEnd - index`` be at the index position of 900, but seeing that the intention was to get the iterator to ``index``, the calculation was off.

Nevertheless,``getSampleFragment`` was still a bit hard to follow. I refactored it once more, and used ``std::distance`` to get the iterator at the correct location instead.